Allow curfile to work in Perl Config files

haukex opened the pull request:
### Summary
Allows `__FILE__`, `__LINE__`, and `Mojo::File::curfile` to work correctly in Perl config files.

### Motivation
Allows one to create filenames relative to the current configuration file in the config. For example:

```
use Mojo::File 'curfile';
{
    migrations_file => curfile->sibling('foo.sql'),
}
```

kraih commented on lib/Mojolicious/Plugin/Config.pm:
```diff
@@ -11,7 +11,7 @@ sub parse {
   # Run Perl code in sandbox
   my $config = eval 'package Mojolicious::Plugin::Config::Sandbox; no warnings;'
-    . "sub app; local *app = sub { \$app }; use Mojo::Base -strict; $content";
+    . "sub app; local *app = sub { \$app }; use Mojo::Base -strict;\n#line 1 $file\n$content";
```
Is that line longer than 80 chars?

haukex replied:
Fixed

Grinnz commented:
I'm not totally against this, but the #line hack can be very finicky and I don't think this is really necessary. It's just as easy to store paths relative to your app home dir in the config file, and have the reader append it to the homedir.

kraih commented on t/mojolicious/perl_config_lite_app.conf:
```diff
@@ -0,0 +1,8 @@
+	utf => "утф",
+	file => curfile->to_string,
+	line => __LINE__, # 7
+}
```
This file doesn't follow our formatting rules.

haukex replied:
If you mean indents as two space chars, I fixed that. Anything else?

jhthorsen replied:
@haukex: You should run `perltidy` on all the files instead of guessing what to change. See #1370 for details, unless you want to install something like https://metacpan.org/pod/App::githook::perltidy

haukex replied:
Thanks, I had missed the `.perltidyrc`.

kraih commented on t/mojolicious/perl_config_lite_app.t:
```diff
@@ -0,0 +1,21 @@
+BEGIN {
+  $ENV{MOJO_MODE}    = 'development';
+  $ENV{MOJO_REACTOR} = 'Mojo::Reactor::Poll';
+}
```
Why is that necessary for the test?

haukex replied:
Cargo-culted from `json_config_lite_app.t`; I removed it.

kraih commented on t/mojolicious/perl_config_lite_app.t:
```diff
@@ -0,0 +1,21 @@
+		utf => "утф",
+		file => curfile->sibling('perl_config_lite_app.conf')->to_string,
+		line => 7,
+	}, 'right value';
```
That also does not look like it follows our formatting rules.

haukex replied:
Fixed as above

haukex commented:
@Grinnz
> the #line hack can be very finicky

Hm, if the filename contains a newline, that might be a problem, so I "fixed" that in the updated commit (not pretty...). Is there anything else?

haukex commented:
I was surprised by the failure on Windows:
```
#          $got->{file} = 'C:/projects/mojo/t/mojolicious/perl_config_lite_app.conf'
#     $expected->{file} = 'C:\projects\mojo\t\mojolicious\perl_config_lite_app.conf'
```
I hope my attempted fix of adding `->to_abs` works.

Grinnz commented:
Spaces in the filename also break the #line directive, but this will just make `__FILE__` not work in that case and not break anything else.

Grinnz commented:
Spaces in a filename work if it's double quoted, but then double quotes in the filename don't work. It's really just not well designed for generated code.

haukex commented:
@Grinnz 
> Spaces in the filename also break the #line directive, but this will just make `__FILE__` not work in that case and not break anything else.

> Spaces in a filename work if it's double quoted, but then double quotes in the filename don't work. It's really just not well designed for generated code.

Hrm, that is indeed not great. On the one hand, I could just add quotes around the filename and hope nothing breaks - but on the other hand, I guess the "defensive" way to code it would be to not include the `#line` directive if it's going to break, so that's what I've done in the latest commit. Still not super pretty, unfortunately...

> It's just as easy to store paths relative to your app home dir in the config file, and have the reader append it to the homedir.

Background for why I suggested this in the first place: I'd like to use the same configuration in a non-Mojo script by `eval`ing it, and it may be in a different path. (And of course adding the `#line` makes error messages nicer.)

kraih commented on lib/Mojolicious/Plugin/Config.pm:
```diff
@@ -9,9 +9,25 @@ sub load { $_[0]->parse(decode('UTF-8', path($_[1])->slurp), @_[1, 2, 3]) }
+      : qq{\n#line 1 "$file"\n}
+      )
+    : qq{\n#line 1 $file\n}
+    );
```
This is really ugly code.

haukex replied:
It's the same code as before, but run through perltidy, as per the Rules in Mojolicious::Guides::Contributing. I could rewrite it as `if` statements for example - what is your preference?

kraih replied:
I don't mean formatting alone. The code also creates lots of untested special cases. Unless the whole feature can be reduced to say one line, i doubt i would vote for it.

haukex replied:
Ah, I see, and yes, [I agree](https://github.com/mojolicious/mojo/pull/1494#issuecomment-615942450). I've made some updates to attempt to address it, [see my comment below](https://github.com/mojolicious/mojo/pull/1494#issuecomment-623005604).

haukex commented:
I looked at Perl's [code](https://github.com/Perl/perl5/blob/8100f3cd3c45f1190105982b7fed1028074526ae/toke.c#L1775) and indeed there is no way to escape double quotes in `#line`, but it looks like any other characters (except `\n`) should work.

So, I've **simplified** the code changes accordingly, added **more tests** to cover the special cases, and added a line in the **documentation**.

I feel like I've done everything I can for now - I hope this passes muster.

kraih commented:
While i might vote in favour of the feature itself if done in the most minimalistic way possible, there is a lot of special cases in this PR i do not agree with. So 👎 from me on this specific implementation of the feature.

kraih commented:
Or to be more precise, there is already `#line` use that should be emulated in `Mojo::Template`.

haukex commented:
@kraih 
Yes, originally it did contain more special cases than necessary (the commits should be squash merged, or I could open a new PR if desired), but now the only thing that's left is `\n` and `"`, just like in `Mojo::Template` - or do you mean other special cases?
I'm not sure I understand your comment in regards to `#line` in `Mojo::Template`, are you suggesting code re-use between the two?

jberger commented:
I personally don't love that the file is evaluated at all and I've been tending to use JSON and YAML config wherever possible. Meanwhile if your case for it is

> Background for why I suggested this in the first place: I'd like to use the same configuration in a non-Mojo script by evaling it, and it may be in a different path.

then I really disagree. My understanding of the reasoning for evaluating was to give access to the app itself. In a non-mojo context, the app having no meaning, you're using a config file as a generic runnable file. If you need that I'd suggest simply doing so, `do $file` the file on your own and load it into `$app->config` at startup. The config loader module isn't magic (apart from running the file), it essentially does the same thing. 👎  from me unfortunately

marcusramberg commented:
I'm also against adding this functionality to the config. 👎

haukex commented:
Thanks for the feedback everyone.

kraih commented:
In case someone stumbles over this in the future, the current solution is:
```
{
    migrations_file => app->home->child('foo.sql'),
}
```
Without the need for imports.
